Sources/MongoSwift/BSON/BSONUtil.swift

```diff
@@ -0,0 +1,15 @@
+internal func getValue(from document: BSONDocument, for key: String) throws -> BSON? {
+    do {
+        return try document.getValue(for: key)
+    } catch let error as BSONError.InvalidArgumentError {
```

Comment:
what about e.g. internal errors?

Reply:
Do you mean emitted from here or in general? I don't think InternalError can be thrown from getValue or setValue
I tried sweeping the driver again it does look like there's one BSONInternalError that get's exposed when we handle boxing and unboxing binary types (specifically UUID) I'll push what I have but I'm unsure how we can 🎣  that

Reply:
it looks like `setValue` can throw an internal error if we exceed the maximum allowed BSON size.
also,`withID()` can also throw if `merge()` fails, for the same reason. we call `withID()` from `BulkWriteOperation` so I think we need to handle that too.

Reply:
Indeed, I missed it, I added the case, but now I realize all the encode methods for each BSONValue can throw BSONError.InternalError. What's the best way to wrap those calls? Is there something I can do with an extension?

Reply:
do you mean the `BSONValue.encode` method implementations? I believe those are only ever called via `setValue` which you are handling, so we are probably ok?

Reply:
Isn't it called when we encode options and such for operations? For example `BulkWriteOperation` has a `BSONEncoder`, won't that call `BSONValue` encode methods?

Reply:
ok so...

there are two `encode` methods on each `BSONValue` type. one of them comes from the `Encodable` protocol and the other from `BSONValue`.

The `Encodable` variants that we implement all just throw an `EncodingError` because right now we don't support using those types with non-BSON encoders. and the `BSONEncoder` itself special-cases all of those types so that the `Encodable.encode` method is never actually called on them.

As for the `BSONValue.encode` implementations...

at a high level what the encoder does is construct an ordered map of `[String: BSONValue]`. once that has been fully filled out, the final step is that we call this method `toDocument()` that (recursively if needed) converts that map to an equivalent `BSONDocument`: https://github.com/mongodb/mongo-swift-driver/blob/ca167f2474524ebbbac0aadb8bfe145e1b0889c8/Sources/MongoSwift/BSON/BSONEncoder.swift#L806

notice that method builds up the document via the subscript setter (which basically calls `getValue` but `fatalError`s on failure). I think this is actually not great, because we'll crash instead of erroring when e.g. the document we create gets too large, so it would probably be wise of us to update that to use `setValue` instead.

in either case, we never directly call the `BSONValue.encode` methods from the encoder. those methods are basically just implementation details of `BSONDocument` and are not at all related to `Codable`. in retrospect I think we should have given this a different name, that's why I went with `write(to buffer: ByteBuffer)` in the basically equivalent method in the new library. 

assuming we make the switch to `setValue` from within the encoder, which means those errors could actually bubble up when encoding a struct now, I think there are two different ways we could handle the driver error conversion: 

1) I'd argue `BSONEncoder` should maybe actually only ever throw `EncodingError`s, and therefore we should wrap the `setValue` call in `toDocument()` in a try/catch that converts any error it sees into an `EncodingError`.
2) alternatively if we do want to propagate BSON errors from the encoder you could extend `BSONEncoder` to add some internal method which does the try/catch, and we use that anywhere in the driver code that we do encoding.

personally I like option 1, what do you think? cc @patrickfreed 

we also have a symmetric decoding problem. the decoder uses `getValue` for retrieving values from `BSONDocument`s. depending on what we decide re: errors, we can either try/catch within the decoder itself and convert those to `DecodingError`s, or add an extension method that does the conversion.

Reply:
Definitely agree on option 1. When I did the error work way back when as part of my rotation, I think I tried to ensure that the encoder/decoder only threw `EncodingError` / `DecodingError`, though I see I may have missed a few cases.